### DISCUSSION TITLE
Broken action file won't appear in PR status checks and so makes the PR green

### DISCUSSION THREAD

**Discussion by Fryguy** [community/community#73048] [Pull Requests]
### Select Topic Area

Bug

### Body

Recently, my team received a PR where a change was made to our ci.yaml workflow.  The change to the workflow was expected, but actually had an invalid syntax change.  Because of that, the workflow failed with a syntax error, however the failure wasn't shown in the PR status checks.  Instead, because the other status checks were green, the PR was entirely shown as green.  We ended up merging the PR because we thought it passed all of the status checks.  Additionally, since it thought all checks were green, it visually collapsed them, so we didn't even notice that the actual CI checks never ran.

Here's the PR for reference: https://github.com/ManageIQ/manageiq/pull/22761
Here's the failed actions run that was not reflected in the PR: https://github.com/ManageIQ/manageiq/actions/runs/6699929362 (broken link, but here's an old screenshot: https://github.com/ManageIQ/manageiq/pull/22761#issuecomment-1787226872)

**Comment by oussamaef06** [comment]
It appears that the issue you described is related to how GitHub handles workflow failures in the context of a pull request. When a workflow fails due to a syntax error or other issues, but other status checks (such as code linting or static analysis) pass, it can sometimes lead to confusion because the overall PR status is shown as "green."

To address this and prevent merging a PR with invalid workflow syntax changes or other issues, consider the following steps:

Review Workflow Changes: Before merging a PR that includes changes to your CI/CD workflow (such as changes to the .github/workflows/ci.yaml file), it's crucial to review these changes thoroughly. Ensure that the workflow syntax is correct and that the workflow will execute as expected.

Testing Workflow Locally: Whenever possible, test the workflow changes locally to catch syntax errors or other issues before pushing the changes to the repository. This can be done using GitHub Actions' [workflow run event](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow) and using a tool like [act](https://github.com/nektos/act) to simulate GitHub Actions locally.

Include Workflow Tests in PR: Consider creating a separate GitHub Actions workflow that checks the validity of your primary CI/CD workflow file. This secondary workflow should run whenever changes are made to the primary workflow. It can check for syntax errors and other issues, and prevent the PR from merging if there are problems.

Status Check Order: Review your repository's settings for branch protection rules and status checks. Ensure that the branch protection rule for your primary branch (e.g., main or master) enforces the requirement for the primary CI/CD workflow to pass before allowing PRs to be merged.

Monitoring Workflow Runs: Pay close attention to the workflow runs for each PR. GitHub provides detailed logs for workflow runs, and you should review these logs to identify any issues, including syntax errors.

Educate Your Team: Make sure your team is aware of the importance of reviewing workflow changes and checking for syntax errors. Encourage a culture of careful review before merging PRs, especially when they involve changes to your CI/CD pipeline.

GitHub Support: If you encounter any issues or inconsistencies with GitHub Actions, consider reaching out to GitHub support for assistance or to report any problems you've encountered.

Incorporating these practices into your workflow can help prevent the merging of PRs with invalid or failing workflows, reducing the risk of accidentally introducing issues into your codebase.

**Reply by Fryguy** [reply]
While some of that advice is useful, that doesn't change the fact that GitHub is not showing failed workflows in the status checks.  None of those workarounds would be needed if the status checks were red.
